Add voltage and amperage scale calibration

IvoFPV opened the pull request:
Adds voltage meter scale calibration by setting the actual battery voltage.
Setting the parameter to anything but zero will adjust scale automatically after saving.

McGiverGim commented:
To me is a good idea, but I don't see it in the middle of the battery tab. I think is better to have a little modal window, that shows this option. Can you look for example to the OSD tab to see how the font manager window is being done to see if it is possible to do something in this way? @mikeller, what do you think?

IvoFPV commented:
I thought about making a separate window like that, but to me it is unnecessary for just one setting and a save button. If there were more settings related to this then it would be a good idea.

mikeller commented on src/js/tabs/power.js:
```diff
@@ -298,6 +300,32 @@ TABS.power.initialize = function (callback) {
+                if (amperagecalibration != 0) {
+                    var amperageoffset = CURRENT_METER_CONFIGS[ampsource - 1].offset / 1000;
+                    if (CURRENT_METERS[ampsource - 1].amperage == amperageoffset) {
+                        amperageoffset += 0.01;
```
Why is this done?

IvoFPV replied:
When scale is high enough, actual amperage will be same as offset, so in the actual calculation that would always cause 0 to be the result.

mikeller replied:
I do not quite get how the scale being high should play into this - if the measured amperage is equal to the offset (i.e. the offset is wrong, or consumption is 0), then obviously it is not possible to calibrate. Instead of just fudging something in this case you should just not calibrate.

IvoFPV replied:
I will work on the modal dialog with all the settings there and update this part as you said, just won't calibrate in that case.

mikeller commented on src/js/tabs/power.js:
```diff
@@ -298,6 +300,32 @@ TABS.power.initialize = function (callback) {
+                        amperageoffset += 0.01;
+                    }
+                    var amperagenewscale = Math.round(CURRENT_METER_CONFIGS[ampsource - 1].scale * 
+                        ((CURRENT_METERS[ampsource - 1].amperage -  amperageoffset) / (amperagecalibration - amperageoffset)));
```
This will cause a division by zero if `amperagecalibration == amperageoffset`.

mikeller commented on src/tabs/power.html:
```diff
@@ -192,6 +192,12 @@
+        <div class="number">
+            <label> <input type="number" name="vbatcalibration" step="0.01" min="0" max="255" /> <span
+                i18n="powerVoltageCalibration"></span>
+                <div class="helpicon cf_tip" i18n_title="powerVoltageCalibrationHelp"></div>
```
There is also a problem with the tooltips stopping to work when the user selects a source that hides them, then selects one that shows them again.

IvoFPV commented:
I just added the modal dialog with its button near Save, to me it looks okay. Maybe I could make the note a bit better to understand.

McGiverGim commented on src/tabs/power.html:
```diff
@@ -207,3 +210,28 @@
     </div>
 </div>
+
+<div class="tab-power" id="calibrationmanagercontent" style="display:none; width:360px;">
```
This div and some of the elements that contain, have a lot of "style" attributes. Please take out them to the CSS. We are trying to not add more of this to the code.

IvoFPV replied:
I removed most of it, so here it is.

mikeller commented on locales/en/messages.json:
```diff
@@ -3383,12 +3383,24 @@
+        "message": "Calibration Manager"
+    },
+    "powerCalibrationManagerHelp": {
+        "message": "<strong>Note:</strong> Before calibrating the scale make sure that divider and multiplier for voltage and offset for amperage is set properly.<br>Setting the values to actual voltage or amperage will calibrate the respective meter scales. Leaving it at 0 will not apply calibration.</br>"
```
I think the message is a bit confusing for a user who does not know what this is about.
I would remove 'Setting the values to actual voltage or amperage will calibrate the respective meter scales.' from the note, and instead put the instructions on top, above the note, like:

'To calibrate, use a multimeter to measure the actual voltage / current draw on your craft (with a battery plugged in), and enter the values below. Then, with the same battery still plugged in, click [Calibrate].'

Also, please add a 'remove props before plugging in a battery' to the note.

IvoFPV replied:
I added the instruction, and prop warning.

mikeller commented on locales/en/messages.json:
```diff
@@ -3383,12 +3383,24 @@
         "message": "$1 V"
     },
+    "powerVoltageCalibration": {
+        "message": "Calibrate Voltage"
```
Probably better 'Measured Voltage', and 'Measured Amperage' below.

mikeller commented on src/js/tabs/power.js:
```diff
@@ -281,6 +297,47 @@ TABS.power.initialize = function (callback) {
+                    }
+                }
+            }
+            save_power_config();
```
This should not be saving the settings - the user might have changed other things in the 'Power' tab that they do not want to keep. Better don't save, but add 'Remember to save your settings after calibrating' to the note in the dialog.

IvoFPV replied:
It only saves scales which are calculated, and no other settings that are changed in power tab are saved. When sources for voltage or current are changed the calibration manager is disabled until the sources are saved.

mikeller replied:
Looking at the code, this is simply not true. It will save **all** settings on the tab, including warning voltages and capacity. The real problem is that it does not give the user any indication that it will do so, potentially overwriting settings that they did not want to overwrite, and it does not allow user to verify that the 'calibrated' settings are actually what they want (they might be obviously out of range, especially for the amperage calibration, where calibrating at idle current without motors running will be extremely prone to be way off).
So, in order to have a good UX, the right thing to do is to present the calibration result to the user as suggestions, and allow them to then decide if they want to save them or not.

IvoFPV replied:
I understand what would be the problem in that case, but I am looking at the code and testing it just now and it is not saving all settings. save_power_config() indeed does save all settings but if they are not modified by parsing the values they stay the same, so only calibration scales get set in CURRENT_METER_CONFIGS and VOLTAGE_METER_CONFIGS. Other settings on tab only get saved when actual save button is pressed and values are parsed.

IvoFPV replied:
Do you think the confirmation dialog of new calculated scales showing the new voltage/current with option to apply or discard them would be a good idea?

mikeller replied:
> I understand what would be the problem in that case, but I am looking at the code and testing it just now and it is not saving all settings. save_power_config() indeed does save all settings but if they are not modified by parsing the values they stay the same,

This is even worse, as this results in buggy behaviour: After closing the calibration dialog, the values displayed to the user in the battery tab will be a mix of saved and unsaved values.

> Do you think the confirmation dialog of new calculated scales showing the new voltage/current with option to apply or discard them would be a good idea?

I think it will be sufficient to allow the user to use the battery tab to confirm the new scale values before saving. Maybe add text to the dialog like 'after calibration, check that the new voltage / amperage scale values are within the expected range, and then hit [Save] to persist them.'

IvoFPV replied:
After closing they get refreshed, meaning they are all saved values. Anyway, I will make a confirmation dialog.

IvoFPV commented:
I added the confirmation dialog showing the new calibrated scales with option to apply or discard them. Applying them will only save the new scale values and not other settings that user might have changed in power tab.

mikeller commented on locales/en/messages.json:
```diff
@@ -3383,12 +3383,39 @@
         "message": "Warning Cell Voltage"
     },
+    "powerCalibrationManager": {
+        "message": "Calibration Manager"
```
'Calibration Manager' is misleading here, as this is not really a 'manager'. I'd just use 'Calibrate'.

mikeller commented on src/js/tabs/power.js:
```diff
@@ -281,6 +306,91 @@ TABS.power.initialize = function (callback) {
+                $('output[name="vbatnewscale"').val(vbatnewscale);
+                $('output[name="amperagenewscale"').val(amperagenewscale);
+
+                $('a.applycalibration').click(function(){
```
This is still wrong - the calibration should not save anything.

IvoFPV commented:
I totally removed saving from the calibration and now when they are applied they only get set as an input value and must be saved by clicking save on power tab.

mikeller commented:
This is starting to look good now.
Two things:
- the 'Calibrate' button is currently only shown if both voltage **and** amperage source support calibration:
![image](https://user-images.githubusercontent.com/4742747/58375695-a07b0f80-7fad-11e9-855e-2d2a2dae5780.png)
Maybe show it if only one supports calibration, and only show the corresponding input field?
- since calibration will not work if the detected voltage or current is not 0, maybe hide the input field for sensors that are reading 0 when opening the dialog as well?

IvoFPV commented:
Calibrate button should be there if there is at least one scale that can be calibrated. When you change voltage or amperage source it gets hidden because that source setting is not saved. After saving it will appear again if any of sources support scaling. 
I will make it to only be possible to calibrate when the readings are not 0, meaning the battery is not plugged in or sources are not set right.

IvoFPV commented:
Now I changed how showing/hiding works. Calibrate button is always shown in bottom bar if there is a scale to calibrate, but when opened it shows warnings if battery not plugged in or sources have changed but haven't been saved. It should work with just voltage or amperage or both. I added checks to remove input if voltage/amperage are less than 0.1, meaning reading is invalid or meter source is incorrect.

IvoFPV commented:
I fixed the whitespace and rebased it.
